jargonLint: initial import of rules

ccoVeille opened the pull request:
these rules are not perfect, they will have to be improved

Source of data and associated license

- https://github.com/sapegin/textlint-rule-terminology MIT cc @sapegin

- https://github.com/huacnlee/autocorrect MIT cc @huacnlee 

- personal data collected and checked via Wikipedia: unsure license

ccoVeille commented:
@alexvonme please review my first version

ccoVeille commented:
Hey @Jason3S I would like you to give me an advice about how to merge my changes.

I don't have much experience as you do with license.

Look at the PR description.

Do you think I should split my PR in multiple one to somehow be able to track from where come each word?

It might be a burden for me because this list is the result of more than 2 years of cherry picking to validate my own markdown.

But I could if needed, please let me know.

Jason3S commented on vale/styles/jargon/TechnologySpecials.yml:
```diff
@@ -0,0 +1,40 @@
+  # manualcherry picking in
+  # https://en.wikipedia.org/wiki/List_of_filename_extensions
+  # https://en.wikipedia.org/wiki/List_of_file_formats
+  (?<=\s)CSON: CSON
```
```suggestion
  \bCSON\b: CSON
```

I suggest using word break `\b` instead of looking for a space. There might not be a space.

Jason3S replied:
Please note: Look behind is not supported by all browsers. Safari only added it last year: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Lookbehind_assertion

ccoVeille replied:
> ```suggestion
>   \bCSON\b: CSON
> ```
> 
> I suggest using word break `\b` instead of looking for a space. There might not be a space.
> 

Great remark I will have a look and check for the use cases I added this pattern for things that could be extensions and terms

But I'm sure it was about handing such use case:

```
JSON is a format, the related extension is .json, example whatever.json 
```

I don't want Vale to report an issue about this valid text.

ccoVeille replied:
> Please note: Look behind is not supported by all browsers. Safari only added it last year: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Lookbehind_assertion

Indeed, but this is a for for vale (a validation tool)

It's in Go, and it supports the syntax.

But I might to keep that in mind because there is a lsp server written in JavaScript for vale

ccoVeille replied:
I think the \b is implicit with Vale, unless I'm playing with `nonword:true`

nonword, bool, Removes the default word boundaries ( \b )

ccoVeille replied:
Anyway, the issue I had was with extension

so here is what I need

```
(?<!\.)CSON: CSON
```

a negative look behind

Jason3S commented on vale/styles/jargon/Suggestions.yml:
```diff
@@ -0,0 +1,20 @@
+---
+extends: substitution
```
A few questions:
- What is the schema for the files?
- Do you need to use Regular Expressions? Would simple conventions work instead?
  - Word breaks are assumed: 
    - start of word: `/(?<=^|\P{L})(?=\p{L})/gmu` (these are not exactly correct)
    - end of word: `/(?<=\p{L})(?=\P{L}|$)/gmu`
  - A space between words will match one or more spaces.
- How do you hope to use these files?

ccoVeille replied:
These are rules for vale tool

https://vale.sh

Jason3S replied:
I'll check it out.

Jason3S commented on vale/styles/jargon/Technology.yml:
```diff
@@ -0,0 +1,326 @@
+  Angular: Angular
+  Ansible: Ansible
+  Apache: Apache
+  API: API
```
These terms seem to be identities. Is there a need to have them in `swap`?

ccoVeille replied:
https://vale.sh/docs/topics/styles/#substitution

I'm using `ignorecsse: true`

So these rules are there to replace any one using

`my api is rich` and reports to use to write `my API is rich`

Jason3S commented:
Copying a word or two from different sources is not generally copyrighted, but taking whole lists of words is.

ccoVeille commented:
Thanks you so much @Jason3S for review this alpha project I asked out of the blue  without much context.

My current work is to provide rules for vale, as it's my need.

Then I may go in a larger direction as described in #1

- #1

I might write something like glosario or frontend-encyclopedia (links are on the quoted discussion)

I expect to create something that could be used to get definition, maybe project URL.

Available as a yaml, JSON, a readme

ccoVeille commented:
To give you a better context of what a vale package look like

You can have a look this rule

https://github.com/errata-ai/packages/blob/master/pkg%2FGoogle%2FGoogle%2FGenderBias.yml

Available in the Google Style rules implemented by Vale

https://vale.sh/hub/google/

ccoVeille commented on vale/styles/jargon/Superlative.yml:
```diff
@@ -0,0 +1,398 @@
+---
```
The two superlative files could/should be moved out of jargon package

More information about this in #11 

- #11 

Thanks @alexvonme for bringing this to my attention

ccoVeille replied:
I killed those files

ccoVeille commented on vale/styles/jargon/TechnologySpecials.yml:
```diff
@@ -0,0 +1,40 @@
+message: Use '%s' instead of '%s'
+level: warning
+ignorecase: true
+scope: raw
```
I found a way to do not use raw, so I will be able to merge these rules with Technology.yml

This was the only reason to split into separate files

alexvonme replied:
I understand, but the reference is a problem. See my comment below.

ccoVeille replied:
I find a way to move them back with the others @Jason3S helped me by suggesting using \b

it helped me to check negative behind rules

alexvonme commented:
@ccoVeille here are my overall thoughts:

- Both superlative files depend on copyrighted content and should be removed.
- TechnologySpecials,yml depends on a Wikipedia article which utilizes CC BY SA which is not a permissive license and therefore the file should also be removed.
- I don't understand the reason you added Suggestions.yml since other organizations already have their (usually larger) versions of that file and wouldn't utilize it.

As such, I'd suggest that you remove all files excluding Technology.yml which we can discuss and expand in this pull request with #6 and #8.

ccoVeille commented on vale/styles/jargon/Technology.yml:
```diff
@@ -0,0 +1,326 @@
+  # yamllint disable rule:line-length # Disable yamlint rule about line length
+  # Sources
+  # - https://github.com/sapegin/textlint-rule-terminology/blob/master/terms.jsonc
+  # - https://github.com/paradite/frontend-encyclopedia/blob/main/scripts/data.json
```
removed

ccoVeille commented:
> @ccoVeille here are my overall thoughts:
> 
>     * Both superlative files depend on copyrighted content and should be removed.

I removed, it was something I had locally that I pushed by errors.
There were files 

>     * TechnologySpecials,yml depends on a Wikipedia article which utilizes CC BY SA which is not a permissive license and therefore the file should also be removed.

I removed them, as I found a way to deal with extension

> 
>     * I don't understand the reason you added Suggestions.yml since it's not an exhaustive list.

This is open for debate, but I can remove them from initial import

> As such, I'd suggest that you remove all files excluding Technology.yml which we can discuss and expand in this pull request with #6 and #8.

It's what I just did

ccoVeille commented:
hey @Jason3S 

my project starts to look like an alpha version of what I imagined, there are still many things to do, but I progressed a lot since I started.

I plan to support export to cspell format #16
- #16 

For now, I can simply copy the term I collected and a open PR on https://github.com/streetsidesoftware/cspell-dicts/tree/main/dictionaries/software-terms

What do you think ?

alexvonme commented on vale/styles/jargonLint/Technology.yml:
```diff
@@ -0,0 +1,325 @@
+  # Sources
+  # - https://github.com/sapegin/textlint-rule-terminology/blob/master/terms.jsonc
+  # - https://github.com/paradite/frontend-encyclopedia/blob/main/scripts/data.json
+  # - https://github.com/huacnlee/autocorrect/blob/main/.autocorrectrc.template
```
Can you separate sources in the Technology.yml file?

As in
```
swap:
  #source1
  a: b
  #source2
  c: d
```

I think that might be better to maintain the list.

ccoVeille replied:
It won't be easy without a master YAML that will refer the source for each word my problem is the fact that if I split and there is let's say JSON in one, and I find it again in another the rules will have 2 sections.

Some may think the issues will be reported because we will have 2 keys with the sane value, but we may not

# section 1
JSON: JSON

# section 2
(?!\.)JSON: JSON

But we can start building the master file

ccoVeille replied:
That's by I opened 

- #16

ccoVeille replied:
Side question, do you think we can merge? 

I mean splitting may require time.

Another option for me, without merging the file is to split the addition to Technology.yaml commit in multiple commits

alexvonme replied:
@ccoVeille if we're not going to clarify where each word came from within the rule, then would it be better to remove sources from the rules entirely? Just creating a sources.md file to keep track for each rule, at least temporarily. 

As it is, I don't think it's a good look. We may keep the rules completely clean and add the comments to that separate file.

alexvonme commented on vale/styles/jargonLint/README.md:
This code will be inaccessible if we separate it from the YAML file. Why not add it there with something like:
```
# You may sort the substitution with "awk -FS=':' '{print $2 $1}' rule.yaml | sort | awk -FS=':' '{print $2 $1}'"
```

alexvonme commented on vale/styles/jargonLint/README.md:
````diff
@@ -0,0 +1,9 @@
+substitution can be sorted with this
+
+```shell
+awk -FS=':' '{print $2 $1}' rule.yaml | sort | awk -FS=':' '{print $2 $1}'
````
This code will be inaccessible if we separate it from the YAML file. Why not add it there with something like:
```
# You may sort the substitution with "awk -FS=':' '{print $2 $1}' rule.yaml | sort | awk -FS=':' '{print $2 $1}'"
```

alexvonme commented on vale/styles/jargonLint/README.md:
```diff
@@ -0,0 +1,9 @@
+
+## Tooling
+
+substitution can be sorted with this
```
Something like: "To sort rules that extend substitution:" might be better.

Instead of extend, you could also use apply or use as well

Jason3S commented:
> hey @Jason3S
> 
> my project starts to look like an alpha version of what I imagined, there are still many things to do, but I progressed a lot since I started.
> 
> I plan to support export to cspell format #16
> 
> * [consider a master YAML files with all the rules and description #16](https://github.com/jargonLint/jargonLint/issues/16)
> 
> For now, I can simply copy the term I collected and a open PR on https://github.com/streetsidesoftware/cspell-dicts/tree/main/dictionaries/software-terms
> 
> What do you think ?

Adding to the Software Terms is fine.

I am thinking about improving file based dictionaries to be more complete. We can talk about it on a different thread.

ccoVeille commented:
That's the purpose of jargonLint I will be more than happy to discuss about it.


I would prefer the discussion to be on jargonLint repository, but I would understand if you prefer to do it on cspell-dicts repository

ccoVeille commented on vale/styles/jargonLint/Technology.yml:
Hey @sagepin,

I saw you fixed your yaml pattern detection rule?

https://github.com/sapegin/textlint-rule-terminology/pull/77#discussion_r1656572029

Did my project/PR inspired you?

ccoVeille replied:
discussion currently in progress about the fact `-` was added to terminology regexp

ccoVeille replied:
Let's merge like the following ticket will address the possible issue

- #24

ccoVeille commented:
@alexvonme pushed back, please review and merge

alexvonme commented:
Reading through it one last time, will merge quickly.

alexvonme commented on vale/styles/jargonLint/Technology.yml:
```diff
@@ -0,0 +1,313 @@
+  RubyGems: RubyGems
+  Rust: Rust
+  SaaS: SaaS
+  Sales?[ -]forces?: Salesforce
```
@ccoVeille does the rule regarding ``Salesforce`` correct ``salesforce`` on your system. It doesn't seem to in mine and I'm not sure why.

alexvonme replied:
Right, just got confused with something else. can you add a question to the middle part, as in ``[ -]?`` so that it also catches ``salesforce`` with the missing capitalization.

ccoVeille replied:
it works because there is `ignorecase:true` I kept the capital S because that's the way to refer to Salesforce

alexvonme replied:
I understood that part. I suggested changes so that ``salesforce`` would also be corrected.  Please go over them when you can.

ccoVeille replied:
I will double-check, because now I get you said the current regexp doesn't while it should

ccoVeille replied:
oh you spotted the error here https://github.com/jargonLint/jargonLint/pull/4#discussion_r1657246078

alexvonme commented on vale/styles/jargonLint/Technology.yml:
```diff
@@ -0,0 +1,313 @@
+  Cloudflare: Cloudflare
+  CloudFront: CloudFront
+  CoffeeScript: CoffeeScript
+  Common[ .]js: CommonJS
```
```suggestion
  Common[ .]?js: CommonJS
```

alexvonme commented on vale/styles/jargonLint/Technology.yml:
```diff
@@ -0,0 +1,313 @@
+  RubyGems: RubyGems
+  Rust: Rust
+  SaaS: SaaS
+  Sales?[ -]forces?: Salesforce
```
```suggestion
  Sales?[ -]?forces?: Salesforce
```

alexvonme commented on vale/styles/jargonLint/Technology.yml:
```diff
@@ -0,0 +1,313 @@
+  Nvidia: Nvidia
+  NYSE: NYSE
+  OAuth: OAuth
+  Objective[\- ]C: Objective-C
```
```suggestion
  Objective[\- ]?C: Objective-C
```

alexvonme commented on vale/styles/jargonLint/Technology.yml:
```diff
@@ -0,0 +1,313 @@
+  RabbitMQ: RabbitMQ
+  React Bootstrap: React Bootstrap
+  React Hook Form: React Hook Form
+  React[ \.]js: React
```
```suggestion
  React[ \.]?js: React
```

ccoVeille commented:
I fixed the one you reported, plus a few one I found by looking at all the file

alexvonme commented:
It might be a good idea to consider removing these 3 commonly used words from the rule:
- rest (listed as REST)
- react (listed as React)
- ok (listed as OK)

ccoVeille commented:
That's right.

We will need a exclusion list to avoid adding them back if someone asks for it.

I mean we should document we don't want them and why.

For now, I'll comment them.

ccoVeille commented:
Fixed
